cpp/src/groupby/groupby.cu

```diff
@@ -129,6 +132,38 @@ groupby::aggregate(std::vector<aggregation_request> const& requests,
   return dispatch_aggregation(requests, 0, mr);
 }
 
+groupby_groups groupby::groups(rmm::mr::device_memory_resource*  mr,
+      cudaStream_t stream) {
+  auto sort_order = helper().key_sort_order(stream);
+  auto group_keys = cudf::experimental::detail::gather(_keys, sort_order);
+  auto group_offsets = helper().group_offsets(stream);
+
+  std::vector<size_type> group_offsets_vector(group_offsets.size());
+
+  thrust::copy(group_offsets.begin(),
+      group_offsets.end(),
+      group_offsets_vector.begin());
+
+  return groupby_groups{std::move(group_keys), group_offsets_vector};
+}
+
+groupby_groups groupby::groups(table_view values, rmm::mr::device_memory_resource*  mr,
+      cudaStream_t stream) {
+  auto sort_order = helper().key_sort_order(stream);
+  auto group_keys = cudf::experimental::detail::gather(_keys, sort_order);
```

Comment:
As I mentioned, the sort order will simply try to put the null keys at the end and then clip them off. You need to slice `sort_order` to new dimensions `[0, helper().num_keys())` and then gather using that.

Since this is using implementation detail unknown to outside world, I'd suggest adding `sorted_keys()` as a sort_helper method.

Reply:
Would `grouped_keys()` be better naming, as the ordering would be congruous with `grouped_values()`?

Reply:
Ah but `grouped_values` already exists.

Reply:
In this case, it' actually sorted keys. A key row in one group is going to be exactly same as rest of the rows in the group. And even then, they are going to be sorted. The reason behind naming for grouped_values was because within a group, the value rows can be in any lexicographical order.

Reply:
Gotcha. Makes sense

Reply:
Done